### DISCUSSION TITLE
SWR and Suspense Future?

### DISCUSSION THREAD

**Discussion by pkellner** [vercel/swr#1944] [General]
I posted this in the NextJS discussion and it got no responses so I'm referencing from here.  I can't find a discussion of how/when Suspense and Next/SWR will be production ready.

https://github.com/vercel/next.js/discussions/36459#discussion-4033755

**Comment by shuding** [comment]
Hey @pkellner – I don't think if we can get a detailed timeline about this as we (Next/SWR/React) need to join forces here. My guess is that we need the [`unstable_getCacheForType`](https://github.com/reactwg/react-18/discussions/25) API or something similar for libraries to complete the Suspense based data fetching story. 

> Ad hoc data fetching with Suspense is technically possible, but still not recommended as a general strategy.

Re-reading this sentence from the React 18 blog post – I'm not sure if it means that it will never be recommended, or there _will_ be an official way to support that strategy better in the future.

**Comment by koba04** [comment]
I'm not sure what the Suspense Future would be. But I guess this includes two topics.

The first one is `unstable_getCacheForType` API that @shuding mentioned. This is required to leverage the power of Concurrent Rendering without de-optimizing back to synchronous. Currently, SWR would have to use `useSyncExternalStore` to be compatible with Concurrent Features. This also introduces Cache Boundary as a new concept. But these tentative APIs seem to be hard to integrate into SWR and other libraries. So it would need more research.

Another is how ad-hoc data libraries, including SWR, encourage the Render As You Fetch pattern. Currently, many developers use the libraries as the Fetch on Render pattern. If they switch to Suspense mode, their applications will cause worse use experiences due to a waterfall problem. Technically, we can avoid this by pre-fetching before rendering. But this is not an intuitive way with the current APIs. I think this is a framework area because it can handle pages and routing and preload asynchronous dependencies before rendering a page. I guess that's the reason the React team recommends using Suspense for Data Fetching with frameworks. I don't think SWR should handle these responsibilities but should provide APIs or guides on how to use Suspense with better user experiences.

**Comment by benevbright** [comment]
> If they switch to Suspense mode, their applications will cause worse use experiences due to a waterfall problem. Technically, we can avoid this by pre-fetching before rendering. But this is not an intuitive way with the current APIs. I think this is a framework area

I'm also wondering if this improvement is in the roadmap of SWR. And also wondering if it's possible for SWR to solve the waterfall issue without requiring preload.
